### PR TITLE
[bitcoin] Revert to berkeley-db@4

### PR DESCRIPTION
**Pull request by casey**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

--

bitcoind requires berkeley db version 4. If only berkeley 5 db is available, support for legacy wallets, which are widely used, will not be available.

The upgrade to berkeley db version 5 probably broke or will break a number of use-cases. I noticed it because Sparrow wallet [broke after upgrading bitcoind](https://github.com/sparrowwallet/sparrow/issues/655).

For reference, see the [MacOS build instructions](https://github.com/bitcoin/bitcoin/blob/master/doc/build-osx.md) and the [unix build instructions](https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md) for bitcoin which mention that berkeley db 4 should be used. (It is possible to use berkeley db 5, but only by passing `--with-incompatible-bdb`, which will result in a bitcoind which is incompatible with the wallets that many people have.

Would it be appropriate to include a note in this PR that berkeley db 4 should not be upgraded without referring to the bitcoind build guides?

This bug was introduced in #107667 by @mhmdanas (no blame! just want to point out that this is a specific, tricky dependency that should not be upgraded.)

### CODE REVIEW

File: Formula/bitcoin.rb [hunk 1]
```diff
@@ -25,7 +25,7 @@ class Bitcoin < Formula
   depends_on "automake" => :build
   depends_on "libtool" => :build
   depends_on "pkg-config" => :build
-  depends_on "berkeley-db@5"
+  depends_on "berkeley-db@4" # do not upgrade
```
Comment (carlocab) [inline Formula/bitcoin.rb:28]: Could we add a link to the documentation here?

Please squash the change into your existing commit and use the commit message
```
bitcoind: revert to berkeley-db@4
```

Thanks!

Comment (casey) [reply Formula/bitcoin.rb]: Definitely! Added and squashed, let me know if that looks good.

File: Formula/bitcoin.rb [hunk 2]
```diff
@@ -25,13 +25,18 @@ class Bitcoin < Formula
...
   depends_on "libevent"
   depends_on macos: :catalina
   depends_on "miniupnpc"
   depends_on "zeromq"
 
+  # berkeley db should be kept at version 4
+  # see the bitcoin build guides for more information:
+  # https://github.com/bitcoin/bitcoin/blob/master/doc/build-osx.md
+  # https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md
+  depends_on "berkeley-db@4"
```
Comment (carlocab) [inline Formula/bitcoin.rb:38]:
```suggestion
  depends_on "boost"
  # berkeley db should be kept at version 4:
  # https://github.com/bitcoin/bitcoin/blob/master/doc/build-osx.md
  # https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md
  depends_on "berkeley-db@4"
  depends_on "libevent"
  depends_on macos: :catalina
  depends_on "miniupnpc"
  depends_on "zeromq"
```
`brew audit` won't pass if dependencies are not listed in alphabetical order. (Squash this into your existing commit again please; thanks!)

**Comment by carlocab** [comment]
We should probably also try to add a test that will fail if `berkeley-db@4` is upgraded to a newer version.

File: Formula/bitcoin.rb (hunk 2 above)
Comment (casey) [reply Formula/bitcoin.rb]: Done!

File: Formula/bitcoin.rb [hunk 3]
```diff
@@ -23,9 +23,12 @@ class Bitcoin < Formula
 
   depends_on "autoconf" => :build
   depends_on "automake" => :build
+  # berkeley db should be kept at version 4
+  # https://github.com/bitcoin/bitcoin/blob/master/doc/build-osx.md
+  # https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md
+  depends_on "berkeley-db@4"
```
Comment (carlocab) [inline Formula/bitcoin.rb:29]: Move this below `depends_on "pkg-config" => :build` and above `depends_on "boost"`.

Comment (casey) [reply Formula/bitcoin.rb:31]: Ah, sorry, I didn't realize it was build dependencies first.

Comment (carlocab) [reply Formula/bitcoin.rb:31]: Sure. No problem. Can we add a test that fails when `berkeley-db@4` is changed to a newer version?

Comment (casey) [reply Formula/bitcoin.rb:31]: I think that's a little bit beyond me, I don't actually know Ruby.

The test would have to:
- Create a tmpdir
- Spawn bitcoind in regtest mode with the tempdir as the data dir. Additionally, you'd have to make sure that the node was configured with ports that wouldn't conflict with a bitcoind instance that the user is running in regtest mode
- Create a legacy wallet (if it fails, bdb wasn't available, or was the wrong version during buidl)
- Delete the tempdir
- Kill the spawned bitcoind instance

Comment (carlocab) [reply Formula/bitcoin.rb:31]: > - Spawn bitcoind in regtest mode with the tempdir as the data dir. Additionally, you'd have to make sure that the node was configured with ports that wouldn't conflict with a bitcoind instance that the user is running in regtest mode
> 
> - Create a legacy wallet (if it fails, bdb wasn't available, or was the wrong version during buidl)

Can you outline what commands you would run in your shell to do these steps? The other three are handled by `brew` mostly automatically.

Comment (casey) [reply Formula/bitcoin.rb:31]: Sure, something like this:

```
TMPDIR=`mktemp -d`
RPCPORT=`freeport`
bitcoind -regtest -rpcport=$RPCPORT -listen=0 -datadir=$TMPDIR &
bitcoin-cli -regtest -datadir=$TMPDIR -rpcport=$RPCPORT createwallet test-wallet false false '' false false
# the above bitcoin-cli command will fail if RPC wallets aren't supported
# kill bitcoind launched above
```

Comment (carlocab) [reply Formula/bitcoin.rb:31]: Ok, I've pushed a change that I think should do that. Please try it out locally to see if it works too. Presumably it should fail with the existing version of `bitcoind`.

Comment (carlocab) [reply Formula/bitcoin.rb:31]: Using the test with `berkeley-db@5` seems to error with `error: Error parsing JSON: ''`. Not sure if that's the error we're looking for?

Comment (carlocab) [reply Formula/bitcoin.rb:31]: Oh, I see. I left out an argument...

Comment (casey) [reply Formula/bitcoin.rb:31]: I believe the argument that's currently `"''"` should just be `""`

Comment (carlocab) [reply Formula/bitcoin.rb:31]: Great, this test seems to work. When built against `berkeley-db@5`, we get:
```
error code: -4
error message:
Compiled without bdb support (required for legacy wallets)
```
https://github.com/carlocab/workflow-test/runs/7946821335?check_suite_focus=true#step:5:251

**Comment by carlocab** [comment]
Ok, test seems to work on ARM. Squashing commits so that this can be merged.

**Review by carlocab — APPROVED** [review]
Thanks!

**Comment by casey** [comment]
Nice! Thank you for taking care of the test!

**Review by carlocab — APPROVED** [review]

**Comment by mhmdanas** [comment]
Ah, my bad, sorry for introducing this bug. At least we have a test for this now.

**Pull request closed without merging**